### PR TITLE
feat(MeetingSdkAdapter): add state for meeting

### PR DESCRIPTION
**Pull request by maloun96**
https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-223892

https://user-images.githubusercontent.com/8277264/114510157-ff4bdf80-9c3e-11eb-98ee-192bcccb5611.mov

### CODE REVIEW

File: src/MeetingsSDKAdapter.js [hunk 1]
```diff
@@ -943,15 +945,30 @@ export default class MeetingsSDKAdapter extends MeetingsAdapter {
...
+      ).pipe(map(() => {
+        const {remoteVideo, state} = this.meetings[ID];
+        const {ENDED, JOINED, ACTIVE} = MeetingState;
+
+        if (state === ENDED) {
+          return this.meetings[ID];
+        }
+
+        if (remoteVideo !== null) {
+          this.meetings[ID].state = JOINED;
```
Comment (cipak) [inline src/MeetingsSDKAdapter.js:957]: remoteVideo not null should mean ACTIVE, right?

File: src/MeetingsSDKAdapter.js [hunk 2]
```diff
@@ -943,15 +942,29 @@ export default class MeetingsSDKAdapter extends MeetingsAdapter {
...
+        const {remoteVideo, state} = this.meetings[ID];
+
+        if (state === MeetingState.ENDED) {
+          return this.meetings[ID];
+        }
+
+        if (remoteVideo !== null) {
+          this.meetings[ID].state = MeetingState.ACTIVE;
+        } else {
+          this.meetings[ID].state = MeetingState.JOINED;
```
Comment (cipak) [inline src/MeetingsSDKAdapter.js:955]: @lalli-flores it seems to me that JOINED is actually the same as ACTIVE, I mean, after you join, you get a different set of meeting controls (eg: Leave meeting). I feel we need to have another talk an these, maybe align them with (or choose a subset of) the sdk meeting states https://github.com/webex/webex-js-sdk/blob/master/packages/node_modules/%40webex/plugin-meetings/src/constants.js#L795

File: src/MeetingsSDKAdapter.test.js [hunk 3]
```diff
@@ -809,8 +809,8 @@ describe('Meetings SDK Adapter', () => {
       meetingSDKAdapter
         .createMeeting(target)
         .pipe(flatMap(() => meetingSDKAdapter.getMeeting(meetingID)))
-        .subscribe((getMeeting) => {
-          expect(getMeeting).toMatchObject(meeting);
+        .toPromise().then((data) => {
+          expect(data).toMatchObject({...meeting, localAudio: null, localVideo: null});
```
Comment (cipak) [inline src/MeetingsSDKAdapter.test.js:813]: why these changes?

Comment (maloun96) [reply src/MeetingsSDKAdapter.test.js:813]: I'm getting this error.
`
Test suite failed to run

    Call retries were exceeded

      at ChildProcessWorker.initialize
`

File: src/MeetingsSDKAdapter.js [hunk 4]
```diff
@@ -936,13 +936,21 @@ export default class MeetingsSDKAdapter extends MeetingsAdapter {
 
       const meetingWithRosterToggleEvent$ = fromEvent(sdkMeeting, EVENT_ROSTER_TOGGLE);
 
+      const meetingStateChange$ = fromEvent(sdkMeeting, EVENT_STATE_CHANGE)
+        .pipe(
```
Comment (cipak) [inline src/MeetingsSDKAdapter.js:940]: pls move this line (pipe) to the line above, like the other observables (line 922, line 926)

File: src/MeetingsSDKAdapter.js [hunk 5]
```diff
@@ -936,13 +936,21 @@ export default class MeetingsSDKAdapter extends MeetingsAdapter {
...
+          tap((event) => {
+            this.meetings[ID].state = event.payload.currentState === 'ACTIVE' ? MeetingState.JOINED : MeetingState.LEFT;
```
Comment (cipak) [inline src/MeetingsSDKAdapter.js:942]: Take care of the 'INACTIVE' state and please create a new meeting object:
```js
const sdkState = event.payloat.currentState;
const meeting = this.meetings[ID];
let state = meeting.state;

if (sdkState === 'ACTIVE') {
  state = MeetingState.JOINED;
}
else if (sdkState === 'INACTIVE') {
  state = ...LEFT;
}

this.meetings[ID] = {...meeting, state};
```

File: src/MeetingsSDKAdapter.js [hunk 6]
```diff
@@ -281,6 +281,7 @@ export default class MeetingsSDKAdapter extends MeetingsAdapter {
       remoteAudio: null,
       remoteVideo: null,
       remoteShare: null,
+      state: MeetingState.LEFT,
```
Comment (cipak) [inline src/MeetingsSDKAdapter.js:284]: not needed anymore (INACTIVE -> LEFT)

**Review by cipak — APPROVED** [review]

File: src/MeetingsSDKAdapter.js [hunk 7]
```diff
@@ -951,7 +967,7 @@ export default class MeetingsSDKAdapter extends MeetingsAdapter {
       this.getMeetingObservables[ID] = getMeetingWithEvents$.pipe(
         publishReplay(1),
         refCount(),
-        takeUntil(end$),
+        takeWhile((meeting) => meeting.state !== MeetingState.LEFT, true),
```
Comment (lalli-flores) [inline src/MeetingsSDKAdapter.js:970]: While we don't expect to get here in our flow, you should include a check for falsy state

```suggestion
        takeWhile((meeting) =>  meeting.state && meeting.state !== MeetingState.LEFT, true),
```

**Review by lalli-flores — COMMENTED** [review]
Minor comments, code looks good otherwise.
Please add a test for the emission when the state event change occurs.

File: src/MeetingsSDKAdapter.js (hunk 7 above)
Comment (maloun96) [reply src/MeetingsSDKAdapter.js]: done

File: src/MeetingsSDKAdapter.js [hunk 8]
```diff
@@ -936,13 +935,30 @@ export default class MeetingsSDKAdapter extends MeetingsAdapter {
...
+          const meeting = this.meetings[ID];
+          let {state} = meeting;
+
+          if (sdkState === 'ACTIVE') {
+            state = MeetingState.JOINED;
+          } else if (sdkState === 'INACTIVE') {
+            state = MeetingState.LEFT;
+          }
+
+          this.meetings[ID] = {...meeting, state};
```
[the remark this answers is not in the document]
Comment (maloun96) [reply src/MeetingsSDKAdapter.js]: done

File: src/MeetingsSDKAdapter.js [hunk 9]
```diff
@@ -936,13 +935,30 @@ export default class MeetingsSDKAdapter extends MeetingsAdapter {
 
       const meetingWithRosterToggleEvent$ = fromEvent(sdkMeeting, EVENT_ROSTER_TOGGLE);
 
+      const meetingStateChange$ = fromEvent(sdkMeeting, EVENT_STATE_CHANGE).pipe(
+        tap((event) => {
+          const sdkState = event.payload.currentState;
+          const meeting = this.meetings[ID];
+          let {state} = meeting;
```
[the remark this answers is not in the document]
Comment (maloun96) [reply src/MeetingsSDKAdapter.js]: done

**Comment by maloun96** [comment]
Hi @lalli-flores. Regarding creating test for emitting state, how exactly ca we test it? This state change comes from plugin-meetings not sdk emits it like **handleLocalAudio**

`sdkMeeting.emit(EVENT_MEDIA_LOCAL_UPDATE, {
        control: AUDIO_CONTROL,
        state: audioEnabled,
      });`

**Comment by lalli-flores** [comment]
@maloun96 because this are unit tests and we're mocking the SDK, we can invoke the event emission at will with a spy, that you can assert in the observable subscription.
I don't want code flows to go untested, but now I'm thinking we should invest in the integration tests instead. I will create some Jiras on this and we can close on this PR for now

**Review by lalli-flores — APPROVED** [review]
